internal/pkg/chain/sampler.go

```diff
@@ -0,0 +1,80 @@
+package chain
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/minio/blake2b-simd"
+
+	"github.com/filecoin-project/go-filecoin/internal/pkg/block"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/crypto"
+)
+
+// A sampler draws randomness seeds from the chain.
+type Sampler struct {
+	reader TipSetProvider
+}
+
+func NewSampler(reader TipSetProvider) *Sampler {
+	return &Sampler{reader}
+}
+
+// Draws a randomness seed from the chain identified by `head` and the highest tipset with height <= `epoch`.
+// If `head` is empty (as when processing the genesis block), the seed is empty.
+func (s *Sampler) Sample(ctx context.Context, head block.TipSetKey, epoch abi.ChainEpoch) (crypto.RandomSeed, error) {
+	var ticket block.Ticket
+	if !head.Empty() {
+		start, err := s.reader.GetTipSet(head)
+		if err != nil {
+			return nil, err
+		}
+		// Note: it is not an error to have epoch > start.Height(); in the case of a run of null blocks the
+		// sought-after height may be after the base (last non-empty) tipset.
+		// It's also not an error for the requested epoch to be negative.
+
+		tip, err := s.findTipsetAtEpoch(ctx, start, epoch)
+		if err != nil {
+			return nil, err
+		}
+		ticket, err = tip.MinTicket()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Sampling for the genesis block.
+		ticket.VRFProof = []byte{}
+	}
+
+	buf := bytes.Buffer{}
+	buf.Write(ticket.VRFProof)
```

Comment:
don't use the `VRFProof` here, use the VRF digest (a hash of the VRF proof)... Annoying I know, but this is how VRFs are defined such that were we to change VRFs the interface would stay the same.

Reply:
Just checking: another blake2 hash is taken just a couple of lines further down (of the buffer including the epoch). Do you still maintain the proof must be hashed separately first?

Reply:
I do alas.

This would make it easier for us to swap out this VRF for another without changing interfaces should we ever choose (Zen can help motivate this if useful)